pkg/causality/internal/slots.go

```diff
@@ -15,53 +15,52 @@ package internal
 
 import (
 	"math"
+	"sort"
 	"sync"
 )
 
-type slot[E SlotNode[E]] struct {
-	nodes map[uint64]E
+type slot struct {
+	nodes map[uint64]*Node
 	mu    sync.Mutex
 }
 
-// SlotNode describes objects that can be compared for equality.
-type SlotNode[T any] interface {
-	// NodeID tells the node's ID.
-	NodeID() int64
-	// Hashs returns the sorted and deduped hashes of the node.
-	Hashes() []uint64
-	// Construct a dependency on `others`.
-	DependOn(dependencyNodes map[int64]T, noDependencyKeyCnt int)
-	// Remove the node itself and notify all dependers.
-	Remove()
-	// Free the node itself and remove it from the graph.
-	Free()
-}
-
 // Slots implements slot-based conflict detection.
-// It holds references to E, which can be used to build
+// It holds references to Node, which can be used to build
 // a DAG of dependency.
-type Slots[E SlotNode[E]] struct {
-	slots    []slot[E]
+type Slots struct {
+	slots    []slot
 	numSlots uint64
 }
 
 // NewSlots creates a new Slots.
-func NewSlots[E SlotNode[E]](numSlots uint64) *Slots[E] {
-	slots := make([]slot[E], numSlots)
+func NewSlots(numSlots uint64) *Slots {
+	slots := make([]slot, numSlots)
 	for i := uint64(0); i < numSlots; i++ {
-		slots[i].nodes = make(map[uint64]E, 8)
+		slots[i].nodes = make(map[uint64]*Node, 8)
 	}
-	return &Slots[E]{
+	return &Slots{
 		slots:    slots,
 		numSlots: numSlots,
 	}
 }
 
+// AllocNode allocates a new node and initializes it with the given hashes.
+// TODO: reuse node if necessary. Currently it's impossible if async-notify is used.
+// The reason is a node can step functions `assignTo`, `Remove`, `free`, then `assignTo`.
+// again. In the last `assignTo`, it can never know whether the node has been reused
+// or not.
+func (s *Slots) AllocNode(hashes []uint64) *Node {
```

Comment:
1. When initializing node, we need to use `sortedDedupKeysHash` to remove duplicates hash values, which is coupled with the internal implementation of slots, but the internal logic of node should not care about this issue.
2. As the holder of node, slots is supposed to be responsible for creating and deleting node. In the future, it will also be responsible for reuse node instances if necessary.